codebase2/util/src/U/Util/Base32Hex.hs

```diff
@@ -1,28 +1,43 @@
+{-# LANGUAGE ViewPatterns #-}
+
 module U.Util.Base32Hex
   ( Base32Hex (..),
     fromByteString,
     toByteString,
+    fromText,
+    validChars,
   )
 where
 
 import Data.ByteString (ByteString)
 import qualified Data.ByteString.Base32.Hex as Base32.Hex
+import Data.Set (Set)
+import qualified Data.Set as Set
 import Data.Text (Text)
 import qualified Data.Text as Text
 import qualified Data.Text.Encoding as Text
 
-newtype Base32Hex = UnsafeBase32Hex {toText :: Text}
+newtype Base32Hex = UnsafeFromText {toText :: Text}
   deriving (Eq, Ord, Show)
 
 -- | Return the lowercase unpadded base32Hex encoding of this 'ByteString'.
 -- Multibase prefix would be 'v', see https://github.com/multiformats/multibase
 fromByteString :: ByteString -> Base32Hex
 fromByteString =
-  UnsafeBase32Hex . Text.toLower . Base32.Hex.encodeBase32Unpadded
+  UnsafeFromText . Text.toLower . Base32.Hex.encodeBase32Unpadded
 
 -- | Produce a 'Hash' from a base32hex-encoded version of its binary representation
 toByteString :: Base32Hex -> ByteString
-toByteString (UnsafeBase32Hex s) =
+toByteString (toText -> s) =
```

Comment:
Nit: To me, this was clearer before (=

When written as a view pattern, it's not totally clear that this is an O(1) pattern match. (I guess because of `PatternSynonyms` a pattern isn't guaranteed to be, either, but most are anyway). But it's not a big deal, the `toText` constructor is right there.

... but also, another nit: since this thing has an unsafe constructor, it probably shouldn't also have a record accessor exported that can be used to change the innards, e.g.

```haskell
myThing {toText = "wob"}
```

So my preference would be something like

```haskell
newtype Base32Hex = UnsafeFromText Text
toText :: Base32Hex -> Text
toText (UnsafeFromText s)  s
```

Reply:
Ah ok, that sounds good.  Record accessor totally not on my radar.